hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala

```diff
@@ -514,6 +514,96 @@ object LowerTableIR {
+          val boundsType = TArray(RVDPartitioner.intervalIRRepresentation(kt))
+          val filterIntervalsRef = Ref(genUID(), boundsType)
+          val filterIntervals: IndexedSeq[Row] = filterPartitioner.rangeBounds.map { i =>
+            RVDPartitioner.intervalToIRRepresentation(i, kt.size)
+          }
+
+          val (newRangeBounds, includedIndices, startAndEndInterval, f) = if (keep) {
+            val (newRangeBounds, includedIndices, startAndEndInterval) = part.rangeBounds.zipWithIndex.flatMap { case (interval, i) =>
+              if (filterPartitioner.overlaps(interval)) {
+                Some((interval, i, (filterPartitioner.lowerBoundInterval(interval).min(nPartitions), filterPartitioner.upperBoundInterval(interval).min(nPartitions))))
+              } else None
+            }.unzip3
+
+            val f: (IR, IR) => IR = {
+              case (partitionIntervals, key) =>
+                // FIXME: don't do a linear scan over intervals. Fine at first to get the plumbing right
+                foldIR(ToStream(partitionIntervals), False()) { case (acc, elt) =>
+                  acc || invoke("partitionIntervalContains",
+                    TBoolean,
+                    elt,
+                    key)
+                }
+            }
+            (newRangeBounds, includedIndices, startAndEndInterval, f)
+          } else {
+            // keep = False
+            val (newRangeBounds, includedIndices, startAndEndInterval) = part.rangeBounds.zipWithIndex.flatMap { case (interval, i) =>
+              val lb = filterPartitioner.lowerBoundInterval(interval)
+              val ub = filterPartitioner.upperBoundInterval(interval)
+              if ((lb until ub).map(filterPartitioner.rangeBounds).exists { filterInterval =>
+                iord.compareNonnull(filterInterval.left, interval.left) <= 0 && iord.compareNonnull(filterInterval.right, interval.right) >= 0
+              })
+                None
+              else Some((interval, i, (filterPartitioner.lowerBoundInterval(interval).min(nPartitions), filterPartitioner.upperBoundInterval(interval).min(nPartitions))))
+            }.unzip3
+
+            val f: (IR, IR) => IR = {
+              case (partitionIntervals, key) =>
+                // FIXME: don't do a linear scan over intervals. Fine at first to get the plumbing right
+                foldIR(ToStream(partitionIntervals), True()) { case (acc, elt) =>
+                  acc && !invoke("partitionIntervalContains",
+                    TBoolean,
+                    elt,
+                    key)
+                }
+            }
+            (newRangeBounds, includedIndices, startAndEndInterval, f)
+          }
+
+          val newPart = new RVDPartitioner(kt, newRangeBounds, part.allowedOverlap)
+
+          TableStage(
+            letBindings = loweredChild.letBindings ++ FastIndexedSeq((filterIntervalsRef.name, Literal(boundsType, filterIntervals))),
```

Comment:
Not a big deal, but `filterIntervalsRef` doesn't need to be in `letBindings`, since it's not used in `contexts` or other `letBindings`. Can just move the definition to `broadcastVals`.

Reply:
thanks, worth fixing.